### PR TITLE
fix(indexer): retry write on PG deadlock

### PR DESCRIPTION
**Pull request by onewland**
Not sure of a clean way to test this, but if people want me to hack around and force a deadlock error I can try to do so

https://getsentry.atlassian.net/browse/SNS-1519

### CODE REVIEW

File: src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py [hunk 1]
```diff
@@ -48,6 +50,41 @@ def _get_db_records(self, use_case_id: UseCaseKey, db_keys: KeyCollection) -> An
...
+        last_seen_exception: Optional[BaseException] = None
+
+        with metrics.timer("sentry_metrics.indexer.pg_bulk_create"):
+            # We use `ignore_conflicts=True` here to avoid race conditions where metric indexer
+            # records might have be created between when we queried in `bulk_record` and the
+            # attempt to create the rows down below.
+            while not success and retry_count + 1 < max_retries:
+                try:
+                    table.objects.bulk_create(new_records, ignore_conflicts=True)
+                    success = True
```
Comment (untitaker) [inline src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py:75]: you can just `break` or `return` here instead of keeping a success flag

**Review by untitaker — COMMENTED** [review]
i would add some sort of exponential `sleep` backoff here, but generally not sure if this is the sanest way to deal with those errors

does the indexer currently skip those messages, or crash entirely?

**Comment by onewland** [comment]
> i would add some sort of exponential `sleep` backoff here, but generally not sure if this is the sanest way to deal with those errors
> 
> does the indexer currently skip those messages, or crash entirely?

I haven't personally verified this but according to the ticket the messages get dropped (not a full crash).

File: src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py (hunk 1 above)
Comment (onewland) [reply src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py:75]: so for the condition on line 83 would you rather I check if retry_count == max_retries - 1?

Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/postgres/postgres_v2.py:75]: if you `return` then you can run the code at L83 unconditionally

**Review by untitaker — COMMENTED** [review]
I am not sure if this is generally a good idea, perhaps check with filippo

**Review by fpacifici — APPROVED** [review]
Please address the comment by Markus

**Pull request merged**